### PR TITLE
Enable alpha channel in color selector

### PR DESCRIPTION
**Pull request by adanielyan**
This pull request adds alpha channel to Color interface in a form of 4th text box.

image

### CODE REVIEW

**Comment by jaycammarano** [comment]
should the bar be labeled? I wouldn't know what that was right now.

**Comment by adanielyan** [comment]
> should the bar be labeled? I wouldn't know what that was right now.

It could be, not sure where it will be the best to put the label though. On the other hand, it should not take a lot of effort to figure out what this slider does, even without a label.

**Comment by adanielyan** [comment]
Does this look better?

image

**Comment by jaycammarano** [comment]
> Does this look better?
> 
> image

I like that.

**Comment by rijkvanzanten** [comment]
> My understanding is that alpha is commonly set as 0 to 100 and stored as rgba(r, g, b, a), though I realize there is also this hex spec.

As a matter of fact, that spec recently got updated, and is now `rgb(r g b a)` where a is 0-100 instead of 0-1 IIRC.

Regardless, opacity should be 0-100, not 0-255 🙂

**Comment by adanielyan** [comment]
> > My understanding is that alpha is commonly set as 0 to 100 and stored as rgba(r, g, b, a), though I realize there is also this hex spec.
> 
> As a matter of fact, that spec recently got updated, and is now `rgb(r g b a)` where a is 0-100 instead of 0-1 IIRC.
> 
> Regardless, opacity should be 0-100, not 0-255 🙂

I think 1-100 is what I am used to as well. The only reason for changing it to 255 was that when converted to hex, 1-100 cannot cover the entire range of 00-FF values while 0-255 can. So 1-100 range will cover only 2/5 of all possible values. That said I can easily change it back to 0-100 if you'd like.

**Comment by adanielyan** [comment]
Is there anything I can do to help move this forward?

**Comment by benhaynes** [comment]
I realize that alpha may be a common use-case, but it does change the saved value output. I'd like to clearly split this out as an interface option: `Enable Alpha Channel` or `Enable Opacity` (toggle), and it should be **off** by default.

* When enabled, the dropdown labels should switch to: `RGBA` and `HSLA` to clarify
* The opacity should just be a fourth INT input directly after the three color options
* Ideally, I'd to have a `%` appended to the opacity input, eg: `56%`. This might require extending the core component to support though... and is not required for this.
* I'd like to fix the system popup, which shows at the top-left of the viewport for me (disconnected from the interface, see below)
* Let's wait on a slider until we can do it right. I'd like to have color+opacity like most design apps, but STILL keep the inputs (see below)

Screen Shot 2021-11-04 at 11 40 23 AM

Screen Shot 2021-11-04 at 11 39 43 AM

I'm happy to do a design for how this should look with _all_ of these features implemented, but that really depends on how far you'd like to take this. If this PR is just about adding opacity, then some of the above is all we need.

**Comment by adanielyan** [comment]
I think the first 2 points are doable. As for the system picker, unfortunately none of the current browsers support alpha channel in the system picker AFAIK. Alternatively we could use a third party picker instead of the system one, but I think it me be too much for such a simple feature.

I'll try to implement the first two recommendations though.

**Comment by benhaynes** [comment]
Awesome! For the system picker, we already have that (if you click on the color in the input), but it shows up in the wrong location within the App. Was just hoping we could get that pinned to the interface.

But that brings up a good point. If the system interface doesn't support opacity, then using that input method wouldn't work if you had "Enable Alpha Channel" on. I wonder how these two inputs will play together. 🤔

**Comment by adanielyan** [comment]
In my implementation that is in this PR, the color picker is only responsible for color, while the opacity is controlled by a separate slider. So the two work nicely together. See the video here: https://github.com/directus/directus/pull/9224#issuecomment-954941983

**Comment by adanielyan** [comment]
I added switch as suggested above.

https://user-images.githubusercontent.com/861777/140844873-79db3c35-e8c1-47a2-87c6-7ceb753d8816.mov

**Comment by adanielyan** [comment]
And the stacked layout

https://user-images.githubusercontent.com/861777/140847867-2ba15abe-701b-4e94-bb9b-1585b1b6d0f5.mov

**Comment by benhaynes** [comment]
Hey @adanielyan — **so** sorry for not being clearer, I meant we should put the "Include Alpha" as an _Interface Option_... not an option in the interface. 😬 

The interface options are set by administrators only (and are written in the interface's `indx.ts` file), so they can control the saved format... which an end user should not be controlling.

Other than that, this is looking really nice!! ❤️

**Comment by adanielyan** [comment]
> Hey @adanielyan — **so** sorry for not being clearer, I meant we should put the "Include Alpha" as an _Interface Option_... not an option in the interface. 😬
> 
> The interface options are set by administrators only (and are written in the interface's `indx.ts` file), so they can control the saved format... which an end user should not be controlling.
> 
> Other than that, this is looking really nice!! ❤️

OK, that's an easy fix.

https://user-images.githubusercontent.com/861777/140975512-0e3a77c5-6f56-466e-b891-79bcc75eb03b.mov

If that's not what you want I give up :)

**Comment by benhaynes** [comment]
Haha, getting closer! But if the Alpha Interface Option (in Settings) is enabled, we don't want/need a second toggle on the interface itself. So let's remove the toggle in the dropdown and make the alpha slider full-width. 👍

**Comment by adanielyan** [comment]
I hope this is the last one :)

https://user-images.githubusercontent.com/861777/141142676-ba4a7908-a62f-4902-920b-1af1ce0516ab.mov

**Comment by benhaynes** [comment]
Perfect, thank you!! The only question now is what should we do with that slider when there is no value (`NULL`). Maybe we conditionally don't show the slider, or use some sort of "default" background color? Thoughts @rijkvanzanten ?

**Comment by adanielyan** [comment]
I thought about that too... Even hid the slider when the value is null, but then the interface looked inconsistent with the slider suddenly appearing after selecting a color, so I reverted it back.

**Comment by benhaynes** [comment]
Yeah, makes sense. The slider is "extra", as we have the input above, so we could simply not include it in the interface. Otherwise, I think we'd need some sort of indeterminate state for the slider. Instead of you going back and forth, I can provide a design (not sure when)... or we can remove the slider for now.

**Comment by benhaynes** [comment]
Yeah, I like that! Better for transparency too. Is that slider background using the global border radius variable? It looks a little squared off... maybe we just make it completely rounded on the ends?

**Comment by adanielyan** [comment]
And in the light mode

https://user-images.githubusercontent.com/861777/141150511-e2fe68da-35bc-41b1-b2f5-c5c8a7e19a7a.mov

**Comment by adanielyan** [comment]
> Yeah, I like that! Better for transparency too. Is that slider background using the global border radius variable? It looks a little squared off... maybe we just make it completely rounded on the ends?

Not sure... I didn't change the border radius of the standard slider.

**Comment by adanielyan** [comment]
It seems it's the border radius of input element, 2px.

**Comment by adanielyan** [comment]
The slider component is broken in Safari. The reason is that the `appearance: none;` is overridden by Safari's `-webkit-appearance: slider-horizontal;`. Adding `-webkit-appearance: none;` in slider's css fixes the problem. However on every commit it is converted to `appearance: none;` which doesn't do the trick. Any ideas how to fix that?

**Comment by benhaynes** [comment]
Not sure about the slider in Safari... I'll leave that to the engineers. But the last (small) request, is to decrease the space between the slider and the color chip options. Looking great, thank you for all this awesome work!

**Comment by rijkvanzanten** [comment]
I think the small border radius is due to the fact that the slider track is made to be 4px high, at which size the border radius is 50%, however, with this thicker track that border radius no longer fits 🙂 

> The reason is that the appearance: none; is overridden by Safari's -webkit-appearance: slider-horizontal;. Adding -webkit-appearance: none; in slider's css fixes the problem. However on every commit it is converted to appearance: none; which doesn't do the trick. Any ideas how to fix that?

Good one.. I believe that's stylelint that's slightly too aggressive. Here's to hoping the Safari team hurries up with the unprefixed support; it's already support in Technology Preview after all! 😄

**Comment by adanielyan** [comment]
Here you go, plus rounded edges

image

image

**Comment by adanielyan** [comment]
> I think the small border radius is due to the fact that the slider track is made to be 4px high, at which size the border radius is 50%, however, with this thicker track that border radius no longer fits 🙂

In fact the rounded corners were added by the browser (Edge). But I made them fully round anyway. I had to add that to v-slider component, but it will not affect the regular slider if `--input-bg-image` css var is not explicitly provided in the slider props.

> Good one.. I believe that's stylelint that's slightly too aggressive. Here's to hoping the Safari team hurries up with the unprefixed support; it's already support in Technology Preview after all! 😄

If it's good for you it's good for me. I don't use Safari anyway! 😄

**Comment by adanielyan** [comment]
Hey @benhaynes, is there anything else blocking this PR? If not, maybe it can be merged? Thanks!

**Comment by benhaynes** [comment]
Just the time it takes to have engineering formally review and merge. We've been abnormally busy on cloud the past few weeks... but should have more time starting next week. 👍

**Comment by adanielyan** [comment]
Sure, understood. Just wanted to make sure there's nothing else to add/fix.

File: app/src/interfaces/select-color/select-color.vue
Comment (rijkvanzanten) [inline app/src/interfaces/select-color/select-color.vue:66]:
```suggestion
			<template v-if="colorType === 'RGB' || colorType === 'RGBA'">
```

File: app/src/interfaces/select-color/select-color.vue
Comment (rijkvanzanten) [inline app/src/interfaces/select-color/select-color.vue:94]:
```suggestion
			<template v-if="colorType === 'HSL' || colorType === 'HSLA'">
```

File: app/src/interfaces/select-color/select-color.vue
Comment (rijkvanzanten) [inline app/src/interfaces/select-color/select-color.vue:137]: Great solution here! One little nitpick, lets rename `--input-bg-image` to `--v-slider-track-background-image`.

File: app/src/interfaces/select-color/select-color.vue
Comment (rijkvanzanten) [inline app/src/interfaces/select-color/select-color.vue:323]: Lets use `padStart` here

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]
Couple of tiny nitpicks / tweaks. Otherwise great addition. Thanks @adanielyan! 😄

**Comment by adanielyan** [comment]
Thanks for suggestions! They have been implemented.

**Comment by adanielyan** [comment]
Hi, is there any other action I need to take before this can be merged? Sorry, I am not very familiar with the process.

**Comment by benhaynes** [comment]
No, you're good. We've just been super busy with other releases and the holidays. We'll try to get this reviewed/merged ASAP.

**Comment by adanielyan** [comment]
Great, thanks @benhaynes!

**Review by rijkvanzanten — APPROVED** [review]

**Pull request merged**
